# Overview of data

This folder is meant to contain the files necessary for analysis, as well as 
store the results generated by the code. The following folders are assumed to exist:

- `bip_schz_data` - containing folders `bip_scz_14`, `bip_scz_18`, `brainvar_results` (which contains `cv_tune_results`), and `gtex_results` (which contains `cv_tune_results`)
- `brainvar_eqtl` 
- `gtex_eqtl` - containing folders `expression_datasets` and `wcgna_results` 
- `gtex_rna_seq`
- `reference_data`
- `bmi` - containing folder `adapt_cv_results`
- `t2d` - containing folder `adapt_cv_results`
- `gene_lists`
- `si_simulations`

The following datasets are __assumed to be in their respective folders__ to start 
the analysis: first for the `bip_schz_data` folder __Schizophrenia (SCZ)__ and __bipolar disorder (BD)__ GWAS results, available to download from the [Psychiatric Genomics Consortium](https://www.med.unc.edu/pgc/download-results/), with the:

- _2014-only_ GWAS datasets saved as `BPSCZ.bp-only.results.txt` and `BPSCZ.scz-only.results.txt`,
- _all 2018_ studies are saved as `BDvsCONT.sumstats.gz` and `SCZvsCONT.sumstats.gz`.

For the `brainvar_eqtl` folder, results are available to download from the [available pre-print for Werling et al. (2019) on bioRxiv](https://www.biorxiv.org/content/10.1101/585430v1):

- BrainVar [eQTLs](https://www.biorxiv.org/content/biorxiv/early/2019/03/22/585430/DC5/embed/media-5.xlsx?download=true), saved as `eqtls.xlsx`, 
- and [WGCNA](https://www.biorxiv.org/content/biorxiv/early/2019/03/22/585430/DC4/embed/media-4.xlsx?download=true), saved as `wgcna.xlsx`.

For the `gtex_eqtl` folder [the Single-Tissue cis-eQTL Data](https://gtexportal.org/home/datasets#filesetFilesDiv55) file corresponding to "eGene and significant variant-gene associations based on permutations" was accessed for identifying GTEx eQTLs and saved in a folder as `GTEx_Analysis_v7_eQTL`

For the `gtex_rna_seq` folder the [RNA-Seq Data](https://gtexportal.org/home/datasets#filesetFilesDiv54) gene TPM counts were used for creating gene co-expression covariates. This includes:

- the list of samples, `GTEx_v7_Annotations_SampleAttributesDS.txt`, 
- a glossary `GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx`, 
- and the TPM counts `GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz`.

For the `reference_data` folder, reference data from the [1000 Genomes reference data](https://ctg.cncr.nl/software/magma) saved as: 

- `g1000_eur.fam`,
- `g1000_eur.bim`, and
- `g1000_eur.bed`.

For the `bmi` folder:

- _2015-only_ BMI studies correspond to the [_GWAS Anthropometric 2015 BMI Summary Statistics_ for EUR ancestry](https://portals.broadinstitute.org/collaboration/giant/images/1/15/SNP_gwas_mc_merge_nogc.tbl.uniq.gz) (Locke et al., 2015), and are saved as `bmi_euro_15.txt`. 
- _all 2018_ BMI studies for replication analysis are from the [updated _BMI and Height GIANT and UK BioBank Meta-analysis Summary Statistics_](https://portals.broadinstitute.org/collaboration/giant/images/c/c8/Meta-analysis_Locke_et_al%2BUKBiobank_2018_UPDATED.txt.gz) (Yengo et al., 2018), and are saved as `bmi_giantuk_18.txt`. 
- WHR studies used as covariates in the BMI analysis are from the [combined EUR ancestry results](https://portals.broadinstitute.org/collaboration/giant/images/5/54/GIANT_2015_WHR_COMBINED_EUR.txt.gz) (Shungin et al., 2015).

Finally, for the `t2d` folder, __Type 2 diabetes (T2D)__ GWAS results are available to download from the [Diabetes Genetics Replication And Meta-analysis (DIAGRAM) consortium](https://www.diagram-consortium.org/downloads.html), under the _T2D GWAS meta-analysis - Unadjusted for BMI_ (Mahajan et al., 2018) and saved as `Mahajan.NatGenet2018b.T2D.European.zip`.

The remaining datasets are all constructed throughout the analysis. The final set
of AdaPT discoveries using gradient boosted trees at target FDR level $\alpha = 0.05$ for the SCZ, BMI, and T2D results are available as `adapt_gwas_results.xlsx`. Sheets contain each unique combination of eSNP and associated cis-eQTL gene.





